Panic during ingestion

A panic happened during ingestion, [log here](https://protocollabs.grafana.net/explore?orgId=1&left=%7B%22datasource%22:%22grafanacloud-logs%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22%7Benvironment%3D%5C%22prod%5C%22,%20namespace%3D%5C%22storetheindex%5C%22,%20pod_name%3D%5C%22indexer-1%5C%22%7D%20%7C%3D%20%5C%22panic%5C%22%22,%22editorMode%22:%22code%22,%22queryType%22:%22range%22%7D%5D,%22range%22:%7B%22from%22:%221664914716271%22,%22to%22:%221664914789729%22%7D%7D). There was not any error logged nearby that is clearly associated.

```
panic: send on closed channel                                                                                                                  
                                                                                                                                               
goroutine 88 [running]:                                                                                                                        
github.com/filecoin-project/storetheindex/internal/ingest.(*Ingester).ingestAd.func6({0x1645dd3?, 0x41e963?}, {{0xc0b5cb3c50?, 0xc00efcd0e0?}}\
, {0x1aac2d0, 0xc00b3ad7a0})                                                                                                                   
        /home/indexer/storetheindex/internal/ingest/linksystem.go:461 +0x333                                                                   
github.com/filecoin-project/go-legs.(*handler).handle.func1({0xc00b4155f0?, 0xc00b583ce0?}, {{0xc0b5cb3c50?, 0x26?}})                          
        /home/indexer/go/pkg/mod/github.com/filecoin-project/go-legs@v0.4.15/subscriber.go:956 +0x116                                          
github.com/filecoin-project/go-legs.wrapBlockHook.func1({0xc00b4155f0, 0x26}, {{0xc0b5cb3c50?, 0x2763c00?}})                                   
        /home/indexer/go/pkg/mod/github.com/filecoin-project/go-legs@v0.4.15/subscriber.go:180 +0xdd                                           
github.com/filecoin-project/go-legs/dtsync.addIncomingBlockHook.func1({0xc00b4155f0, 0x26}, {0x1ab6340, 0xc087392000}, {0x1ab4cf0?, 0xc0b0eb6b\
10?}, {0x1ab1208, 0xc0b0eb6b40})                                                                                                               
        /home/indexer/go/pkg/mod/github.com/filecoin-project/go-legs@v0.4.15/dtsync/sync.go:144 +0xa2                                          
github.com/filecoin-project/go-legs/dtsync.(*Sync).addRateLimiting.func1({0xc00b4155f0, 0x26}, {0x1ab6340, 0xc087392000}, {0x1ab4cf0, 0xc0b0eb\
6b10}, {0x1ab1208?, 0xc0b0eb6b40?})                                                                                                            
        /home/indexer/go/pkg/mod/github.com/filecoin-project/go-legs@v0.4.15/dtsync/sync.go:137 +0x10e                                         
github.com/ipfs/go-graphsync/requestmanager/hooks.blockHookDispatcher({0x151d4e0?, 0xc087392040?}, {0x1443980?, 0xc00010bfe0?})                
        /home/indexer/go/pkg/mod/github.com/ipfs/go-graphsync@v0.13.2/requestmanager/hooks/blockhooks.go:25 +0x11b                             
github.com/hannahhoward/go-pubsub.(*PubSub).Publish(0xc00a4f2ec0, {0x151d4e0, 0xc087392040})                                                   
        /home/indexer/go/pkg/mod/github.com/hannahhoward/go-pubsub@v0.0.0-20200423002714-8d62886cc36e/pubsub.go:76 +0x12d                      
github.com/ipfs/go-graphsync/requestmanager/hooks.(*IncomingBlockHooks).ProcessBlockHooks(0xc000014a30, {0xc00b4155f0, 0x26}, {0x1ab6340, 0xc0\
87392000}, {0x1ab4cf0, 0xc0b0eb6b10})                                                                                                          
        /home/indexer/go/pkg/mod/github.com/ipfs/go-graphsync@v0.13.2/requestmanager/hooks/blockhooks.go:42 +0x132                             
github.com/ipfs/go-graphsync/requestmanager/executor.(*Executor).processBlockHooks(0xc0003f3220, {0xc00b4155f0, 0x26}, {0x1ab6340, 0xc08739200\
0}, {0x1ab4cf0?, 0xc0b0eb6b10?})                                                                                                               
        /home/indexer/go/pkg/mod/github.com/ipfs/go-graphsync@v0.13.2/requestmanager/executor/executor.go:162 +0x72                            
github.com/ipfs/go-graphsync/requestmanager/executor.(*Executor).onNewBlock(0x10c03a?, {{0x1ab47b0, 0xc00b4debc0}, {0x1abf978, 0xc00d0f6550}, \
{{{0xc00acf8630, 0x25}}, {0x1ac7240, 0xc00d2ac440}, 0x0, ...}, ...}, ...)                                                                      
        /home/indexer/go/pkg/mod/github.com/ipfs/go-graphsync@v0.13.2/requestmanager/executor/executor.go:172 +0x19c                           
github.com/ipfs/go-graphsync/requestmanager/executor.(*Executor).processResult(0xc0003f3220?, {{0x1ab47b0, 0xc00b4debc0}, {0x1abf978, 0xc00d0f\
6550}, {{{0xc00acf8630, 0x25}}, {0x1ac7240, 0xc00d2ac440}, 0x0, ...}, ...}, ...)                                                               
        /home/indexer/go/pkg/mod/github.com/ipfs/go-graphsync@v0.13.2/requestmanager/executor/executor.go:201 +0x16b                           
github.com/ipfs/go-graphsync/requestmanager/executor.(*Executor).traverse(0x0?, {{0x1ab47b0, 0xc00b4debc0}, {0x1abf978, 0xc00d0f6550}, {{{0xc0\
0acf8630, 0x25}}, {0x1ac7240, 0xc00d2ac440}, 0x0, ...}, ...})                                                                                  
        /home/indexer/go/pkg/mod/github.com/ipfs/go-graphsync@v0.13.2/requestmanager/executor/executor.go:154 +0x525                           
github.com/ipfs/go-graphsync/requestmanager/executor.(*Executor).ExecuteTask(0xc0003f3220, {0x1ab47b0, 0xc00a4f31c0}, {0xc00b4155f0, 0x26}, 0x\
c00095eb01?)                                                                                                                                   
        /home/indexer/go/pkg/mod/github.com/ipfs/go-graphsync@v0.13.2/requestmanager/executor/executor.go:83 +0x566                            
github.com/ipfs/go-graphsync/taskqueue.(*WorkerTaskQueue).worker(0xc00b588320, {0x1aa6ec0, 0xc0003f3220})                                      
        /home/indexer/go/pkg/mod/github.com/ipfs/go-graphsync@v0.13.2/taskqueue/taskqueue.go:139 +0x3b5                                        
created by github.com/ipfs/go-graphsync/taskqueue.(*WorkerTaskQueue).Startup                                                                   
        /home/indexer/go/pkg/mod/github.com/ipfs/go-graphsync@v0.13.2/taskqueue/taskqueue.go:97 +0x3a                                          
indexer@ip-172-31-40-238:storetheindex> git branch                                                                                             
  concurrent-ingest
  ```

--- a/command/daemon.go
+++ b/command/daemon.go
@@ -621,7 +621,6 @@ func reloadConfig(cfgPath string, ingester *ingest.Ingester, reg *registry.Regis
 		}
 		ingester.SetBatchSize(cfg.Ingest.StoreBatchSize)
 		ingester.RunWorkers(cfg.Ingest.IngestWorkerCount)
-		ingester.SetWriteEntriesSynchronously(cfg.Ingest.WriteEntriesSynchronously)
 	}
 
 	err = setLoggingConfig(cfg.Logging)

--- a/config/ingest.go
+++ b/config/ingest.go
@@ -70,11 +70,6 @@ type Ingest struct {
 	// or a chain of advertisement entries. The value is an integer string
 	// ending in "s", "m", "h" for seconds. minutes, hours.
 	SyncTimeout Duration
-	// Synchronously, when true, tells the indexer to process entry chunks
-	// synchronously, waiting for each to complete before fetching the next.
-	// Otherwise, the indexer processes entry chunks asynchronously. This value
-	// is updated when the configuration is reloaded.
-	WriteEntriesSynchronously bool
 }
 
 // NewIngest returns Ingest with values set to their defaults.

--- a/doc/config.md
+++ b/doc/config.md
@@ -93,8 +93,7 @@ config file at runtime.
     "ResendDirectAnnounce": true,
     "StoreBatchSize": 4096,
     "SyncSegmentDepthLimit": 2000,
-    "SyncTimeout": "2h0m0s",
-    "WriteEntriesSynchronously": false
+    "SyncTimeout": "2h0m0s"
   },
   "Logging": {
     "Level": "info",
@@ -232,8 +231,7 @@ Default:
   "ResendDirectAnnounce": false,
   "StoreBatchSize": 4096,
   "SyncSegmentDepthLimit": 2000,
-  "SyncTimeout": "2h0m0s",
-  "WriteEntriesSynchronously": false
+  "SyncTimeout": "2h0m0s"
 }
 ```
 
@@ -286,6 +284,5 @@ The storetheindex daemon can reload some portions of its config without restarti
 - [`Ingest.IngestWorkerCount`](#ingest)
 - [`Ingest.RateLimit`](#ingestratelimit)
 - [`Ingest.StoreBatchSize`](#ingest)
-- [`Ingest.WriteEntriesSynchronously`](#ingest)
 - [`Logging`](#logging)
 - [`Peering`](#peering)

--- a/internal/ingest/ingest.go
+++ b/internal/ingest/ingest.go
@@ -137,11 +137,6 @@ type Ingester struct {
 
 	// Multihash minimum length
 	minKeyLen int
-
-	// writeEntsSync is accessed through SetWriteEntriesSynchronously() and
-	// writeEntriesSynchronously(). It tells the Ingester to handle writing entry chunks
-	// synchronously, waiting for each to complete before fetching the next.
-	writeEntsSync uint32
 }
 
 // NewIngester creates a new Ingester that uses a go-legs Subscriber to handle
@@ -216,7 +211,6 @@ func NewIngester(cfg config.Ingest, h host.Host, idxr indexer.Interface, reg *re
 
 	ing.workersCtx, ing.cancelWorkers = context.WithCancel(context.Background())
 	ing.RunWorkers(cfg.IngestWorkerCount)
-	ing.SetWriteEntriesSynchronously(cfg.WriteEntriesSynchronously)
 
 	go ing.runIngesterLoop()
 
@@ -263,20 +257,6 @@ func (ing *Ingester) getRateLimiter(publisher peer.ID) *rate.Limiter {
 	}
 	// Return rate limiter with rate setting from config.
 	return rate.NewLimiter(ing.rateLimit, ing.rateBurst)
-}
-
-// SetWriteEntriesSynchronously, when set to true, tells the indexer to process chunks
-// of multihash entries synchronously.
-func (ing *Ingester) SetWriteEntriesSynchronously(val bool) {
-	var b32 uint32
-	if val {
-		b32 = 1
-	}
-	atomic.StoreUint32(&ing.writeEntsSync, b32)
-}
-
-func (ing *Ingester) writeEntriesSynchronously() bool {
-	return atomic.LoadUint32(&ing.writeEntsSync) != 0
 }
 
 func (ing *Ingester) Close() error {

--- a/internal/ingest/linksystem.go
+++ b/internal/ingest/linksystem.go
@@ -390,22 +390,6 @@ func (ing *Ingester) ingestAd(publisherID peer.ID, adCid cid.Cid, ad schema.Adve
 	} else {
 		log = log.With("entriesKind", "EntryChunk")
 
-		var chunkFuncs chan func()
-		var asyncDone chan struct{}
-		var asyncEntries bool
-
-		if !ing.writeEntriesSynchronously() {
-			asyncEntries = true
-			asyncDone = make(chan struct{})
-			chunkFuncs = make(chan func(), 1)
-			go func(fch <-chan func()) {
-				for f := range fch {
-					f()
-				}
-				close(asyncDone)
-			}(chunkFuncs)
-		}
-
 		// We have already peeked at the first EntryChunk as part of probing
 		// the entries type, so process that first.
 		chunk, err := ing.loadEntryChunk(syncedFirstEntryCid)
@@ -413,27 +397,15 @@ func (ing *Ingester) ingestAd(publisherID peer.ID, adCid cid.Cid, ad schema.Adve
 			errsIngestingEntryChunks = append(errsIngestingEntryChunks, err)
 		} else {
 			chunkStart := time.Now()
-			if asyncEntries {
-				chunkFuncs <- func() {
-					if err := ing.ingestEntryChunk(ctx, ad, syncedFirstEntryCid, *chunk, log); err != nil {
-						errsIngestingEntryChunks = append(errsIngestingEntryChunks, err)
-					}
-				}
-			} else {
-				err = ing.ingestEntryChunk(ctx, ad, syncedFirstEntryCid, *chunk, log)
-				if err != nil {
-					errsIngestingEntryChunks = append(errsIngestingEntryChunks, err)
-				}
+			err = ing.ingestEntryChunk(ctx, ad, syncedFirstEntryCid, *chunk, log)
+			if err != nil {
+				errsIngestingEntryChunks = append(errsIngestingEntryChunks, err)
 			}
 			mhCount += len(chunk.Entries)
 			entsStoreElapsed += time.Since(chunkStart)
 		}
 
 		if chunk != nil && chunk.Next != nil {
-			var errCh chan error
-			if asyncEntries {
-				errCh = make(chan error, 1)
-			}
 			nextChunkCid := chunk.Next.(cidlink.Link).Cid
 			// Traverse remaining entry chunks based on the entries selector
 			// that limits recursion depth.
@@ -448,33 +420,11 @@ func (ing *Ingester) ingestAd(publisherID peer.ID, adCid cid.Cid, ad schema.Adve
 					errsIngestingEntryChunks = append(errsIngestingEntryChunks, err)
 					return
 				}
-				if asyncEntries {
-					// If an error occurred, cause the segment sync to fail
-					// with that error.
-					select {
-					default:
-					case err = <-errCh:
-						actions.FailSync(err)
-						return
-					}
-					chnk := *chunk
-					chunkFuncs <- func() {
-						if err := ing.ingestEntryChunk(ctx, ad, c, chnk, log); err != nil {
-							errsIngestingEntryChunks = append(errsIngestingEntryChunks, err)
-							// Tell segment sync to fail with the err.
-							select {
-							case errCh <- err:
-							default:
-							}
-						}
-					}
-				} else {
-					err = ing.ingestEntryChunk(ctx, ad, c, *chunk, log)
-					if err != nil {
-						actions.FailSync(err)
-						errsIngestingEntryChunks = append(errsIngestingEntryChunks, err)
-						return
-					}
+				err = ing.ingestEntryChunk(ctx, ad, c, *chunk, log)
+				if err != nil {
+					actions.FailSync(err)
+					errsIngestingEntryChunks = append(errsIngestingEntryChunks, err)
+					return
 				}
 				mhCount += len(chunk.Entries)
 				entsStoreElapsed += time.Since(chunkStart)
@@ -486,20 +436,12 @@ func (ing *Ingester) ingestAd(publisherID peer.ID, adCid cid.Cid, ad schema.Adve
 				}
 			}))
 			if err != nil {
-				if asyncEntries {
-					close(chunkFuncs)
-					<-asyncDone
-				}
 				wrappedErr := fmt.Errorf("failed to sync entries: %w", err)
 				if strings.Contains(err.Error(), "content not found") {
 					return adIngestError{adIngestContentNotFound, wrappedErr}
 				}
 				return adIngestError{adIngestSyncEntriesErr, wrappedErr}
 			}
-		}
-		if asyncEntries {
-			close(chunkFuncs)
-			<-asyncDone
 		}
 	}
 	ing.signalMetricsUpdate()

